### PR TITLE
New staking mechanics

### PR DESCRIPTION
**Pull request by hysz**
## Description

This PR introduces a new mechanism for managing stake and tracking rewards. It is more flexible than the original implementation, allowing for a better user experience.

Note - Skip the Typescript in this review. There is a separate PR (#2126) that updates client-side code and adds tests. It will be merged into this PR before merging into `3.0`.

*Design Principles*

1. Freshly minted stake is active.
2. Delegating, un-delegating and re-delegating stake comes into effect next epoch.
3. Users can freely adjust the distribution of their stake for the next epoch.
4. Stake can be withdrawn after it is inactive for one epoch.

## Stake State Algorithm

There are three states that stake can exist in: Active, Inactive or Delegated. Each state has three fields:

1. How much stake is currently in this state (cur)
2. How much stake is in this state next epoch (next)
3. The last time this state was stored
    1. This allows us to compute the correct values at any given epoch without user intervention

Inactive stake includes a Withdrawable field (W) that reflects how much stake can be withdrawn at any given time.

*Example*

StakingWithWithdrawals (3) (1)

## Reward Tracking Mechanics

Screen Shot 2019-08-30 at 8 16 33 PM

## Reward Vault vs Eth Vault

The reward vault tracks the balance of each pool: how much belongs to the operator and how much is to be split between the delegators. ETH is deposited into this vault when an epoch is finalized.

The ETH Vault stores balances on a per-address basis. Value is moved from the reward vault to the ETH vault when a delegator changes their stake in a given pool.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol [hunk 1]
```diff
@@ -17,297 +17,225 @@
...
-        address payable member,
-        uint256 amountOfStakeToUndelegate,
-        uint256 totalStakeDelegatedToPoolByMember,
-        uint256 totalStakeDelegatedToPool
+        uint256 memberStakeOverInterval,
+        uint256 beginEpoch,
+        uint256 endEpoch
     )
-        internal
+        private
```
Comment (jalextowle) [inline contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol]: I understand why this is being made private, but it may be wise to leave this as `internal` for the purposes of unit testing. Thoughts?

File: contracts/staking/contracts/src/stake/MixinZrxVault.sol [hunk 2]
```diff
@@ -45,4 +48,47 @@ contract MixinZrxVault is
...
     }
+
+    /// @dev Deposits Zrx Tokens from the `owner` into the vault.
+    /// @param owner of Zrx Tokens
+    /// @param amount of tokens to deposit.
+    function _depositFromOwnerIntoZrxVault(address owner, uint256 amount)
+        internal
+    {
+        IZrxVault _zrxVault = zrxVault;
+        require(
```
Comment (jalextowle) [inline contracts/staking/contracts/src/stake/MixinZrxVault.sol]: Is there a reason not to use Rich Errors in this contract?

File: contracts/staking/contracts/src/stake/MixinStakeStorage.sol [hunk 3]
```diff
@@ -0,0 +1,196 @@
+/*
+
+  Copyright 2018 ZeroEx Intl.
```
Comment (jalextowle) [inline contracts/staking/contracts/src/stake/MixinStakeStorage.sol]: Nit:

```suggestion
  Copyright 2019 ZeroEx Intl.
```

File: contracts/staking/contracts/src/stake/MixinStakeStorage.sol [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        balancePtr.lastStored = balance.lastStored;
+        balancePtr.next = balance.next;
+        balancePtr.current = balance.current;
+    }
+
+    /// @dev Returns true iff storage pointers resolve to same storage location.
+    /// @param balancePtrA first storage pointer.
+    /// @param balancePtrB second storage pointer.
+    /// @return true iff pointers are equal.
+    function _arePointersEqual(
```
Comment (jalextowle) [inline contracts/staking/contracts/src/stake/MixinStakeStorage.sol]: Nice!

File: contracts/staking/contracts/src/stake/MixinStakeStorage.sol [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    /// @dev Stores a balance in storage.
+    /// @param balancePtr points to where `balance` will be stored.
+    /// @param balance to save to storage.
+    function _storeBalance(
+        IStructs.DelayedBalance storage balancePtr,
+        IStructs.DelayedBalance memory balance
+    )
+        private
```
Comment (jalextowle) [inline contracts/staking/contracts/src/stake/MixinStakeStorage.sol]: Nit: Same as above re testing

File: contracts/staking/contracts/src/fees/MixinExchangeFees.sol [hunk 6]
```diff
@@ -218,9 +223,20 @@ contract MixinExchangeFees is
             );
 
             // record reward in vault
-            _recordDepositInStakingPoolRewardVault(activePools[i].poolId, reward);
+            bool rewardForOperatorOnly = activePools[i].delegatedStake == 0;
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/fees/MixinExchangeFees.sol]: Ah I actually thought that Solidity would stop managing stack variables once they were finished being used without needing an explicit scope ~ just verified this is not the case. I'm cool with inlining.

File: contracts/staking/contracts/src/interfaces/IStakingPoolRewardVault.sol [hunk 7]
```diff
@@ -75,44 +81,57 @@ interface IStakingPoolRewardVault {
...
     /// Note that this is only callable by the staking contract, and when
     /// not in catastrophic failure mode.
     /// @param poolId Unique Id of pool.
     /// @param amount Amount in ETH to record.
-    function recordDepositFor(bytes32 poolId, uint256 amount)
-        external;
+    /// @param operatorOnly Only attribute amount to operator.
+    /// @return operatorPortion Portion of amount attributed to the operator.
+    /// @return operatorPortion Portion of amount attributed to the delegators.
+    function recordDepositFor(
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/interfaces/IStakingPoolRewardVault.sol]: Ah this is a function signature but the `return` does look weird... I'll play with the indentation.

File: contracts/staking/test/utils/staking_wrapper.ts [hunk 8]
```diff
@@ -97,7 +104,7 @@ export class StakingWrapper {
         this._validateDeployedOrThrow();
         return this._LibFeeMathTestContractIfExists as LibFeeMathTestContract;
     }
-    public async deployAndConfigureContractsAsync(): Promise<void> {
+    public async deployAndConfigureContractsAsync(customStakingArtifact?: any): Promise<void> {
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/test/utils/staking_wrapper.ts]: Hmm uncertain. It feels kind of hacky. We're experimenting with other techniques for the unit tests. I could see us pushing some pattern that achieves this upstream to the wrappers.

File: contracts/staking/test/actors/delegator_actor.ts [hunk 9]
```diff
@@ -8,6 +8,7 @@ import { DelegatorBalances, StakerBalances } from '../utils/types';
 import { StakerActor } from './staker_actor';
 
 export class DelegatorActor extends StakerActor {
+    /**
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/test/actors/delegator_actor.ts]: (please see #2126)

File: contracts/staking/test/utils/staking_wrapper.ts [hunk 10]
```diff
@@ -419,57 +377,52 @@ export class StakingWrapper {
...
         logUtils.log(`Finalization costed ${txReceipt.gasUsed} gas`);
         return txReceipt;
     }
-    public async skipToNextEpochAsync(): Promise<TransactionReceiptWithDecodedLogs> {
-        // increase timestamp of next block
-        const epochDurationInSeconds = await this.getEpochDurationInSecondsAsync();
-        await this._web3Wrapper.increaseTimeAsync(epochDurationInSeconds.toNumber());
-        // mine next block
+    public async fastForwardToNextEpochAsync(): Promise<void> {
+         // increase timestamp of next block
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/test/utils/staking_wrapper.ts]: We'll merge #2126, which has the typescript updates, into this branch before merging into `3.0`.

File: contracts/staking/test/utils/Simulation.ts [hunk 11]
```diff
@@ -15,6 +15,9 @@ chaiSetup.configure();
 const expect = chai.expect;
 
 export class Simulation {
+    /*
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/test/utils/Simulation.ts]: See #2126 for updated tests. We'll focus more on the simulation after the implementation / unit tests are finished.

File: contracts/staking/contracts/src/vaults/EthVault.sol [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    /// @dev Withdraw ALL ETH to `msg.sender` from the vault.
+    function withdrawAll()
+        external
+        returns (uint256)
+    {
+        // get total balance
+        address payable owner = msg.sender;
+        uint256 totalBalance = balances[owner];
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/vaults/EthVault.sol]: Counter Nit ~ we tend to favor explicit returns opposed to implicit 😬 . But, I do agree we could switch to a named return value :)

File: contracts/staking/contracts/src/stake/MixinStake.sol [hunk 13]
```diff
@@ -17,188 +17,217 @@
...
 import "../staking_pools/MixinStakingPoolRewardVault.sol";
+import "../staking_pools/MixinStakingPoolRewards.sol";
 import "../sys/MixinScheduler.sol";
 import "./MixinStakeBalances.sol";
-import "./MixinTimeLockedStake.sol";
+import "./MixinStakeStorage.sol";
 
 
 /// @dev This mixin contains logic for managing ZRX tokens and Stake.
-/// Stake is minted when ZRX is deposited and burned when ZRX is withdrawn.
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStake.sol]: That's a good question. I was also thinking about the best way to do this. The spec that I've written so far is in markdown. I'm thinking that we'll just put it right here in the repo. And then it can have indices that map to the code. So if you're looking at a mixin like 🤔 ~ then you'd hit up the spec and search for the Mixin. We'll mention this in the README too.

File: contracts/staking/contracts/src/stake/MixinStake.sol [hunk 14]
```diff
@@ -17,188 +17,217 @@
...
-    /// @param amount of Zrx to deposit / Stake to mint.
-    function depositZrxAndMintDeactivatedStake(uint256 amount)
+    /// @dev Stake ZRX tokens. Tokens are deposited into the ZRX Vault. Unstake to retrieve the ZRX.
+    ///      Stake is in the 'Active' state.
+    /// @param amount of ZRX to stake.
+    function stake(uint256 amount)
         external
     {
-        _mintStake(msg.sender, amount);
+        address payable owner = msg.sender;
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStake.sol]: Hmm I hadn't thought of that. How about `staker`?

File: contracts/staking/contracts/src/stake/MixinStake.sol [hunk 15]
```diff
@@ -17,188 +17,217 @@
...
-        stakeByOwner[owner] = stakeByOwner[owner].safeAdd(amount);
+        // sync cumulative rewards that we'll need for future computations
+        _syncCumulativeRewardsNeededByDelegator(poolId, currentEpoch);
 
-        // emit stake event
-        emit StakeMinted(
-            owner,
-            amount
-        );
+        // decrement how much stake the owner has delegated to the input pool
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStake.sol]: *facepalm* 🙏

File: contracts/staking/contracts/src/vaults/StakingPoolRewardVault.sol [hunk 16]
```diff
@@ -222,21 +252,35 @@ contract StakingPoolRewardVault is
...
 
-        // update balances
-        balance.operatorBalance = uint256(balance.operatorBalance).safeAdd(operatorPortion).downcastToUint96();
-        balance.membersBalance = uint256(balance.membersBalance).safeAdd(poolPortion).downcastToUint96();
+        // compute new balances
+        uint256 newOperatorBalance = uint256(balance.operatorBalance).safeAdd(operatorPortion);
+        uint256 newMembersBalance = uint256(balance.membersBalance).safeAdd(poolPortion);
+
+        // save new balances
+        balance.operatorBalance = LibSafeDowncast.downcastToUint96(newOperatorBalance);
```
Comment (abandeali1) [inline contracts/staking/contracts/src/vaults/StakingPoolRewardVault.sol]: We can use `newOperatorBalance.downcastToUint96` here

File: contracts/staking/contracts/src/vaults/EthVault.sol [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    Authorizable,
+    IEthVault,
+    IVaultCore,
+    MixinVaultCore
+{
+
+    using LibSafeMath for uint256;
+
+    // mapping from Owner to ETH balance
+    mapping (address => uint256) internal balances;
```
Comment (abandeali1) [inline contracts/staking/contracts/src/vaults/EthVault.sol]: Nit: `_balances`

File: contracts/staking/contracts/src/vaults/EthVault.sol [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+pragma solidity ^0.5.9;
+
+import "@0x/contracts-utils/contracts/src/LibSafeMath.sol";
+import "../interfaces/IEthVault.sol";
+import "./MixinVaultCore.sol";
+
+
+/// @dev This vault manages ETH.
+contract EthVault is
+    Authorizable,
```
Comment (abandeali1) [inline contracts/staking/contracts/src/vaults/EthVault.sol]: All of these are unnecessary exccept `IEthVault` and `MixinVaultCore`

File: contracts/staking/contracts/src/vaults/EthVault.sol [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+        // notify
+        emit EthDepositedIntoVault(msg.sender, owner, amount);
+    }
+
+    /// @dev Withdraw an `amount` of ETH to `msg.sender` from the vault.
+    /// Note that only the Staking contract can call this.
+    /// Note that this can only be called when *not* in Catostrophic Failure mode.
+    /// @param amount of ETH to withdraw.
+    function withdraw(uint256 amount)
```
Comment (abandeali1) [inline contracts/staking/contracts/src/vaults/EthVault.sol]: A lot of these functions are exactly the same as in the ZrxVault. Maybe the repeat functions should live in `MixinVaultCore`?

File: contracts/staking/contracts/src/sys/MixinScheduler.sol [hunk 20]
```diff
@@ -36,6 +36,7 @@ import "../interfaces/IStakingEvents.sol";
 contract MixinScheduler is
     IStakingEvents,
     MixinDeploymentConstants,
+    Ownable,
```
Comment (abandeali1) [inline contracts/staking/contracts/src/sys/MixinScheduler.sol]: This can be removed.

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol [hunk 21]
```diff
@@ -17,297 +17,225 @@
...
-import "../libs/LibStakingRichErrors.sol";
-import "../libs/LibRewardMath.sol";
 import "../immutable/MixinStorage.sol";
 import "../immutable/MixinConstants.sol";
 import "../stake/MixinStakeBalances.sol";
 import "./MixinStakingPoolRewardVault.sol";
-import "./MixinStakingPool.sol";
 
 
-/// @dev This mixin contains logic for staking pool rewards.
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol]: (see above)

File: contracts/staking/contracts/src/stake/MixinStakeStorage.sol (hunk 5 above)
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStakeStorage.sol]: Yeah, I think you're right. For unit testing these will have to be made `internal` =\

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol (hunk 1 above)
Comment (hysz) [reply contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol]: Yeah, I think you're right. For unit testing these will have to be made internal =\

File: contracts/staking/contracts/test/TestStorageLayout.sol [hunk 22]
```diff
@@ -25,6 +25,9 @@ import "../src/interfaces/IStructs.sol";
 
 
 contract TestStorageLayout is
+    MixinDeploymentConstants,
+    Ownable,
+    MixinConstants,
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/test/TestStorageLayout.sol]: Yeah, to remove any ambiguities during linearization it will print out the full ancestral tree. Ugly, but it works. I'll create a task to remove frivolous imports for readability to see if we can make this a bit prettier. In the meantime I'll cleanup the ones in this PR.

[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/test/TestStorageLayout.sol]: FWIW this is an example demonstrating why explicitly inheriting the ancestral tree works:

```
Ex:
contract A {}
contract B {}
contract C {}
contract D is A, B {}

Suppose now that there is a contract E that will inherit from B{} and D{}
The following follows "most base-like" to "most-derived":
contract E is B, D {}
However, it unfolds to: B{}, A{}, B{}, D{}, E{} - which is wrong because B{} appears both before and after A{}.
The correct contract inheritance for E{} should be:
contract E is A{}, B{}, D{}

So, even though contract E{} does not directly depend on A{} - it still must be included in the dependency list.
```

File: contracts/staking/contracts/src/vaults/StakingPoolRewardVault.sol [hunk 23]
```diff
@@ -50,6 +52,9 @@ contract StakingPoolRewardVault is
     // mapping from Pool to Reward Balance in ETH
     mapping (bytes32 => Balance) internal balanceByPoolId;
 
+    // address of ether vault
+    IEthVault internal ethVault;
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/vaults/StakingPoolRewardVault.sol]: Ah my bad, I didn't realize we were prefixing with `_` for internal state variables. This will have to change throughout the staking contracts. I'm going to leave it consistent for now but created a task to update all of our internal state variables.

File: contracts/utils/contracts/src/LibSafeMath.sol [hunk 24]
```diff
@@ -6,6 +6,8 @@ import "./LibSafeMathRichErrors.sol";
 
 library LibSafeMath {
 
+    using LibSafeMath for uint256;
```
Comment (abandeali1) [inline contracts/utils/contracts/src/LibSafeMath.sol]: I don't think this is used anywhere.

File: contracts/staking/contracts/test/LibFeeMathTest.sol [hunk 25]
```diff
@@ -2,7 +2,7 @@
 
 /*
 
-  Copyright 2018 ZeroEx Intl.
```
Comment (abandeali1) [inline contracts/staking/contracts/test/LibFeeMathTest.sol]: Do we still need this file at all?

File: contracts/staking/contracts/src/immutable/MixinStorage.sol [hunk 26]
```diff
@@ -40,26 +41,23 @@ contract MixinStorage is
     // address of staking contract
     address internal stakingContract;
 
-    // mapping from Owner to Amount Staked
-    mapping (address => uint256) internal stakeByOwner;
+    // mapping from Owner to Amount of Active Stake
+    mapping (address => IStructs.DelayedBalance) internal activeStakeByOwner;
```
Comment (abandeali1) [inline contracts/staking/contracts/src/immutable/MixinStorage.sol]: Nit: all of these variables should be prefixed with `_`. Probably cleaner to do this in another PR.

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol [hunk 27]
```diff
@@ -17,297 +17,223 @@
...
-
-        return amount;
+        // cache some values to reduce sloads
+        IStructs.DelayedBalance memory delegatedStake = delegatedStakeToPoolByOwner[member][poolId];
+        uint256 currentEpoch = getCurrentEpoch();
+
+        // value is always zero in these two scenarios:
+        //   1. The current epoch is zero: delegation begins at epoch 1
+        //   2. The owner's delegated is current as of this epoch: their rewards have been moved to the ETH vault.
+        if (currentEpoch == 0 || delegatedStake.lastStored == currentEpoch) return 0;
```
Comment (abandeali1) [inline contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol]: These checks should probably be reversed, since we don't need `currentEpoch == 0` to be checked after epoch 0.

File: contracts/staking/contracts/src/vaults/EthVault.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (jalextowle) [reply contracts/staking/contracts/src/vaults/EthVault.sol]: Fair enough. I was counter nitted on this a few other times yesterday, so I'll try to update my internal style guide 👍

File: contracts/staking/contracts/src/interfaces/IStakingPoolRewardVault.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (jalextowle) [reply contracts/staking/contracts/src/interfaces/IStakingPoolRewardVault.sol]: Oops, you're totally right

File: contracts/staking/contracts/src/stake/MixinStake.sol [hunk 28]
```diff
@@ -17,188 +17,227 @@
...
 {
     using LibSafeMath for uint256;
 
-    /// @dev Deposit Zrx. This mints stake for the sender that is in the "Deactivated & Withdrawable" state.
-    /// @param amount of Zrx to deposit / Stake to mint.
-    function depositZrxAndMintDeactivatedStake(uint256 amount)
+    /// @dev Stake ZRX tokens. Tokens are deposited into the ZRX Vault. Unstake to retrieve the ZRX.
+    ///      Stake is in the 'Active' state.
+    /// @param amount of ZRX to stake.
+    function stake(uint256 amount)
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStake.sol]: Note: we chatted about this offline, but basically this won't dilute rewards because the stake isn't delegated.

File: contracts/staking/contracts/src/interfaces/IStructs.sol [hunk 29]
```diff
@@ -56,15 +56,44 @@ interface IStructs {
...
+    /// @dev Balance struct for stake.
+    /// @param current Balance in the current epoch.
+    /// @param next Balance in the next epoch.
+    struct StakeBalance {
+        uint256 current;
+        uint256 next;
+    }
+
+    /// @dev States that stake can exist in.
+    enum StakeState {
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/interfaces/IStructs.sol]: Hmm yeah I like that better too. Although a "status machine" doesn't quite have the same ring as a "state machine" 🌝

File: contracts/staking/contracts/src/stake/MixinStakeStorage.sol [hunk 30]
```diff
@@ -0,0 +1,199 @@
...
+    {
+        // note - this compresses into a single `sstore` when optimizations are enabled,
+        // since the StakeBalance struct occupies a single word of storage.
+        balancePtr.lastStored = balance.lastStored;
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStakeStorage.sol]: Yeah unfortunately that assignment won't work in Solidity. Because `balancePtr` is a storage pointer, the `=` operator changes what it points to but not the underlying value.

File: contracts/staking/contracts/test/LibFeeMathTest.sol (hunk 25 above)
Comment (hysz) [reply contracts/staking/contracts/test/LibFeeMathTest.sol]: It's removed by #2109

File: contracts/staking/contracts/src/immutable/MixinStorage.sol (hunk 26 above)
Comment (hysz) [reply contracts/staking/contracts/src/immutable/MixinStorage.sol]: Agreed, I've created a task for this.

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol (hunk 27 above)
Comment (hysz) [reply contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol]: Very good eye! I agree!

File: contracts/staking/contracts/src/stake/MixinStake.sol (hunk 14 above)
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStake.sol]: Yeah we're going to need to change this in a lot of places. This name is used all throughout the codebase. I've created a task to do this once all outstanding PR's are merged, to avoid all the conflicts.

File: contracts/staking/contracts/src/stake/MixinStakeBalances.sol [hunk 31]
```diff
@@ -24,165 +24,139 @@ import "../interfaces/IStructs.sol";
...
-    {
-        return totalActivatedStake;
-    }
-
     /// @dev Returns the total stake for a given owner.
-    /// This stake can be in any state.
-    /// @param owner to query.
+    /// @param owner of stake.
     /// @return Total active stake for owner.
     function getTotalStake(address owner)
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStakeBalances.sol]: This will be useful for the end-user to query exactly how much ZRX they have staked.

[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStakeBalances.sol]: They could go directly to the ZRX Vault, but I'd prefer to have a single point-of-entry for queries. Additionally, they don't really need to be exposed to the vault unless there's a catastrophic failure.

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewardVault.sol [hunk 32]
```diff
@@ -98,39 +64,43 @@ contract MixinStakingPoolRewardVault is
...
-    }
-
     /// @dev Deposits an amount in ETH into the reward vault.
     /// @param amount The amount in ETH to deposit.
     function _depositIntoStakingPoolRewardVault(uint256 amount)
         internal
     {
+        // cast to payable and sanity check
         address payable rewardVaultAddress = address(uint160(address(rewardVault)));
+        if (rewardVaultAddress == NIL_ADDRESS) {
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewardVault.sol]: Note: we chatted about this offline. If we decide to make these changes they will be in a separate PR.

File: contracts/staking/contracts/src/vaults/EthVault.sol (hunk 17 above)
Comment (hysz) [reply contracts/staking/contracts/src/vaults/EthVault.sol]: Note: we chatted offline. This will be done in a separate PR.

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol [hunk 33]
```diff
@@ -17,297 +17,222 @@
...
+        // cache a storage pointer to the cumulative rewards for `poolId` indexed by epoch.
+        mapping (uint256 => IStructs.Fraction) storage cumulativeRewardsByPoolPtr = cumulativeRewardsByPool[poolId];
+
+        // fetch the last epoch at which we stored an entry for this pool;
+        // this is the most up-to-date cumulative rewards for this pool.
+        uint256 cumulativeRewardsLastStored = cumulativeRewardsByPoolLastStored[poolId];
+        IStructs.Fraction memory mostRecentCumulativeRewards = cumulativeRewardsByPoolPtr[cumulativeRewardsLastStored];
+
+        // compute new cumulative reward
+        (uint256 numerator, uint256 denominator) = LibFractions.addFractions(
```
Comment (jalextowle) [inline contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol]: Would it make more sense for `addFractions` to return a fraction? It seems like we create a Fraction object a significant amount of the time after calling this.

File: contracts/staking/contracts/src/stake/MixinStake.sol [hunk 34]
```diff
@@ -17,188 +17,227 @@
...
+    /// @dev Delegates a owners stake to a staking pool.
+    /// @param poolId Id of pool to delegate to.
+    /// @param owner who wants to delegate.
+    /// @param amount of stake to delegate.
+    function _delegateStake(
+        bytes32 poolId,
+        address payable owner,
+        uint256 amount
+    )
+        private
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStake.sol]: Note: we chatted about this offline. These will be set to `internal` on as needed, unless a better alternative arises in the meantime.

File: contracts/staking/contracts/src/stake/MixinStakeStorage.sol [hunk 35]
```diff
@@ -0,0 +1,199 @@
...
+
+
+/// @dev This mixin contains logic for managing stake storage.
+contract MixinStakeStorage is
+    IStakingEvents,
+    MixinConstants,
+    MixinStorage,
+    MixinScheduler
+{
+
```
[the remark this answers is not in the document]
Comment (hysz) [reply contracts/staking/contracts/src/stake/MixinStakeStorage.sol]: Interesting, yeah I think you're right. Seems to be 50/50 whether we have a newline here or not lol.

File: contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol (hunk 33 above)
Comment (hysz) [reply contracts/staking/contracts/src/staking_pools/MixinStakingPoolRewards.sol]: Mm idk, since this is a general util library I'd prefer to not force a struct on future users. We actually only convert it to the Fraction struct in this fn for storage.

File: contracts/staking/contracts/src/vaults/EthVault.sol [hunk 36]
```diff
@@ -0,0 +1,108 @@
...
+    /// Note that only the Staking contract can call this.
+    /// Note that this can only be called when *not* in Catostrophic Failure mode.
+    /// @param owner of ETH Tokens.
+    function depositFor(address owner)
+        external
+        payable
+    {
+        // update balance
+        uint256 amount = msg.value;
+        balances[owner] = balances[owner].safeAdd(msg.value);
```
Comment (jalextowle) [inline contracts/staking/contracts/src/vaults/EthVault.sol]: We might as well use `amount` here since `msg.value` was cached there, right? This said, `CALLVALUE` is typically actually cheaper than using a stack variable (it's a Gbase opcode), so I'd be in favor of just using `msg.value` for both uses and removing `amount` entirely, unless there is a readability concern.

File: contracts/staking/contracts/src/vaults/EthVault.sol [hunk 37]
```diff
@@ -0,0 +1,108 @@
...
+    /// @dev Withdraw an `amount` of ETH to `owner` from the vault.
+    /// @param owner of ETH.
+    /// @param amount of ETH to withdraw.
+    function _withdrawFrom(address payable owner, uint256 amount)
+        internal
+    {
+        // update balance
+        // note that this call will revert if trying to withdraw more
+        // than the current balance
+        balances[owner] = balances[owner].safeSub(amount);
```
Comment (jalextowle) [inline contracts/staking/contracts/src/vaults/EthVault.sol]: Nice reentrancy protection 👍
